source/adapters/level_zero/program.cpp

```diff
@@ -234,8 +249,17 @@ UR_APIEXPORT ur_result_t UR_APICALL urProgramCompile(
   // This produces better code because the driver can do cross-module
   // optimizations.  Therefore, we just remember the compilation flags, so we
   // can use them later.
-  if (Options)
+  if (Options) {
     Program->BuildFlags = Options;
+
+    // if large allocations are selected, then pass
+    // ze-opt-greater-than-4GB-buffer-required to disable
+    // stateful optimizations and be able to use larger than
+    // 4GB allocations on these kernels.
+    if (Context->Devices[0]->useOptimized32bitAccess() == 0) {
```

Comment:
Should this be in the Exp function and have the Exp function have the Compile code now? That way you check the specific device being passed in and not just the Context->Devices[0] since you might be on a non-uniform system.

Reply:
thanks @nrspruit . Good idea. However, `urProgramCompileExp` at this moment is unimplemented, so adding implementation for `urProgramCompileExp` on top of these changes would make this PR too big. I think it is better we merge this patch, then we add the support for `urProgramCompileExp`, including using the functionality from this patch. what do you think?

Reply:
Sure, that would be fine, a follow-up patch would be good improvement on this.